library/alloc/src/vec/mod.rs

```diff
@@ -2032,18 +2032,41 @@ impl<T, A: Allocator> Vec<T, A> {
     #[stable(feature = "rust1", since = "1.0.0")]
     #[track_caller]
     pub fn insert(&mut self, index: usize, element: T) {
-        #[cold]
-        #[cfg_attr(not(feature = "panic_immediate_abort"), inline(never))]
-        #[track_caller]
-        #[optimize(size)]
-        fn assert_failed(index: usize, len: usize) -> ! {
-            panic!("insertion index (is {index}) should be <= len (is {len})");
-        }
+        let _ = self.insert_mut(index, element);
+    }
 
+    /// Inserts an element at position `index` within the vector, shifting all
+    /// elements after it to the right.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `index > len`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(push_mut)]
+    /// let mut vec = vec![1, 3, 5, 9];
+    /// let x = vec.insert_mut(3, 6);
+    /// *x += 1;
+    /// assert_eq!(vec, [1, 3, 5, 7, 9]);
+    /// ```
+    ///
+    /// # Time complexity
+    ///
+    /// Takes *O*([`Vec::len`]) time. All items after the insertion index must be
+    /// shifted to the right. In the worst case, all elements are shifted when
+    /// the insertion index is 0.
+    #[cfg(not(no_global_oom_handling))]
+    #[inline]
+    #[unstable(feature = "push_mut", issue = "135974")]
+    #[track_caller]
+    #[must_use = "if you don't need a reference to the value, use Vec::insert instead"]
+    pub fn insert_mut(&mut self, index: usize, element: T) -> &mut T {
         let len = self.len();
-        if index > len {
-            assert_failed(index, len);
-        }
+        if intrinsics::unlikely(index > len) {
```

Comment:
Why this change? `unlikely` is a relatively weak hint compared to explicit outlining, and it seems unrelated to this PR.

Reply:
My bad. I got lazy.

Reply:
@ibraheemdev Fixed.

Reply:
Actually, come to think of it, this code almost exactly is in VecDeque. Should this be abstracted out, or left as is??

Reply:
Keep the call to `assert_failed` here, and the `assert!` in `vec_deque`. `panic!` and similar are already tuned to assume that they won't be hit.

(Using `likely`/`unlikely` should effectively always be justified by benchmarks anyway, those hints can make codegen worse if not used properly).